### ISSUE TITLE
[nu-validator coverage] <img> inside <figure> without <figcaption> should require alt

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#3944]
# [nu-validator coverage] `<img>` inside `<figure>` without `<figcaption>` should require `alt`

## Repro

```html
<figure>
  <img src="image.jpg">
</figure>
```

nu-validator: error — "An 'img' element must have an 'alt' attribute, except under certain conditions."
markuplint (all rules enabled by the bench): 0 violations.

Fixture: `tests/external/validator/tests/html/elements/img/missing-alt-in-figure-novalid.html` (`nu-only`).

## Spec

HTML LS "Requirements for providing text to act as an alternative for images" (`images.html#alt`) and its subsection "Guidance for conformance checkers" (`images.html#guidance-for-conformance-checkers`) enumerate when `alt` is required and the narrow exceptions under which it may be omitted. The `<figure>` + `<figcaption>` case is treated in "A key part of the content" (`images.html#a-key-part-of-the-content`).

Anchors are stable; the sub-section numbering (currently in the 4.8.4.4.x range) shifts between drafts and is omitted.

Applied to the fixture: no `<figcaption>` is present, so none of the omission conditions apply — `alt` is required.

## Current markuplint behaviour

- `required-attr` requires user-supplied `nodeRule` config to enforce `alt` on `img`; nothing spec-driven fires by default (see `packages/@markuplint/rules/src/required-attr/index.spec.ts` `[required-attr-invalid-001]` — the test wraps the check in an explicit `nodeRule`).
- `packages/@markuplint/html-spec/src/spec.img.jsonc` defines `alt` as `{ "type": "Any" }` without a `required` flag or `requiredEither` (unlike `src` / `srcset`).
- The conditional shape ("required unless one of the exceptions in `images.html#guidance-for-conformance-checkers` applies") is not expressible with the current `requiredEither` grammar.

## Proposal

Either:

1. Extend `required-attr` (or add a new spec-driven rule, e.g. `img-alt-required`) that encodes the exception list from `images.html#guidance-for-conformance-checkers`. See the spec for the full and authoritative list of exemptions — do not enumerate it in the rule's user docs; link to the spec anchor.

2. Add a richer `condition` grammar to `requiredEither` so `spec.img.jsonc` can express the exception structure declaratively.

The `wai-aria-*` family and `require-accessible-name` do not cover this: the fixture's `<img>` has no explicit role, no `aria-label`, no `aria-labelledby`, and neither rule fires.

## Priority

Deferred.

## Implementation notes

- A new default-on `alt`-required rule (or `required-attr` extension) will fire on many real-world pages that markuplint currently accepts silently — the implementation PR must include release-note wording and consider whether the default severity should be `warning` or `error`.
- This Issue is registered as the `primary` mapping for the fixture in `tests/external/bench/issue-xref.config.ts`; xref refresh (`yarn bench:xref`) is the source of truth for the current fixture set backing this Issue.

## Benchmark cross-reference

_Sync'd from `tests/external/snapshots/` (submodule=1429313, nu-validator@sha256:210bbc353b23, markuplint=5.0.0-rc.4, generated=2026-08-12 03:50:09 UTC)._

Filter: `^html\/elements\/img\/missing-alt-in-figure-novalid` → **1** fixtures. Verdict tally: match-error=0, match-clean=0, ml-only=0, nu-only=1, nu-over=0.

[HTML LS "Requirements for providing text to act as an alternative for images"](https://html.spec.whatwg.org/multipage/images.html#alt) and its "Guidance for conformance checkers" subsection ([`images.html#guidance-for-conformance-checkers`](https://html.spec.whatwg.org/multipage/images.html#guidance-for-conformance-checkers)) enumerate the narrow exceptions under which `alt` may be omitted. The fixture has no `<figcaption>` inside the enclosing `<figure>`, so none of the omission conditions apply. `packages/@markuplint/rules/src/required-attr/index.spec.ts` `[required-attr-invalid-001]` demonstrates that markuplint's `required-attr` fires on `img` only via explicit `nodeRule` config; `packages/@markuplint/html-spec/src/spec.img.jsonc` defines `alt` as `{ "type": "Any" }` without a `required` flag or `requiredEither`. The current `requiredEither` grammar cannot express the exception-list shape from the spec — a new rule (working name `img-alt-required`) or a richer condition grammar is required.

| fixture | verdict | nu id / ml rule |
| --- | --- | --- |
| `html/elements/img/missing-alt-in-figure-novalid.html` | nu-only | `nv-224e80c23676` |

Reproduce: see [`bench-triage` skill](https://github.com/markuplint/markuplint/blob/dev/.claude/skills/bench-triage/SKILL.md) for the audit workflow.

**Comment by YusukeHirao** [comment]
Duplicate triage landed on this same fixture independently (filed as #3964, closing in favor of this Issue). One factual correction to the "Current markuplint behaviour" section:

> The `wai-aria-*` family and `require-accessible-name` do not cover this: the fixture's `<img>` has no explicit role, no `aria-label`, no `aria-labelledby`, and neither rule fires.

Verified empirically against the actual fixture (`tests/external/validator/tests/html/elements/img/missing-alt-in-figure-novalid.html`) via `mlRuleTest` — `require-accessible-name` **does** fire:

```json
[{ "severity": "error", "message": "Require accessible name", "line": 4, "col": 3, "raw": "<img src=\"image.jpg\">" }]
```

Role `img` has `accessibleNameRequired: true`, and `getAccessibleName()`'s `handleImg` (`packages/@markuplint/ml-spec/src/algorithm/aria/accname/element-names.ts`) returns `null` when both `alt` and `title` are absent, which `require-accessible-name` treats as a violation.

The real reason this stays `nu-only` isn't rule-logic absence — it's that `require-accessible-name` ships only via `markuplint:a11y` (`specConformance: "non-normative"`, an ARIA-heuristic preset), and `tests/external/bench/config.ts` intentionally does not extend that preset (bench scope is HTML LS / WAI-ARIA / URL LS conformance, not accessibility best-practice heuristics — see `.claude/skills/bench-triage/SKILL.md`'s "ml-only readings" classification for the general principle).

This doesn't change the proposal (a dedicated normative rule mirroring HTML LS §4.8.4.4's default case is still the bench-aligned path — a11y-preset heuristics stay out of scope for the bench by design), but it's worth noting `require-accessible-name` is functionally adequate for real-world usage today; the gap is purely a bench/preset-scoping one, not an implementation one.

**Comment by YusukeHirao** [comment]
Resolved without a new rule. `require-accessible-name` (already shipped, category `a11y`) computes `accessibleNameRequired` from the ARIA role spec — the `img` role's WAI-ARIA 1.2 definition states "authors MUST provide a label using the aria-label or aria-labelledby attribute," and `alt` is the HTML-side accessible-name source for `img`. So an `<img>` with no `alt` (regardless of ancestor context, e.g. inside `<figure>`) already produces a violation from this rule — verified directly against the fixture:

```
$ markuplint --config '{"rules":{"require-accessible-name":true}}' missing-alt-in-figure-novalid.html
<markuplint> error: Require accessible name (require-accessible-name) ...
```

The gap was purely that `tests/external/bench/config.ts` (the benchmark's curated rule subset) never enabled `require-accessible-name`, so this nu-only fixture never got a chance to match. Added `'require-accessible-name': true` there; the fixture now flips `nu-only` → `match-error`.

The original proposal's premise — that `required-attr`/`spec.img.jsonc`'s `alt` field can't express the omission-exception shape, so a new `img-alt-required` rule or richer condition grammar is needed — undersold what `require-accessible-name` already covers generically (any role with `accessibleNameRequired: true`, not just `img`). No code change to `@markuplint/rules` or `@markuplint/html-spec` was needed.

Closing as resolved by the bench config change (`tests/external/bench/config.ts`).
